### PR TITLE
limiting domain names to 255/63 octets/labels

### PR DESCRIPTION
**Pull request by ogud**
title says it all

### CODE REVIEW

File: msg.go [hunk 1]
```diff
@@ -369,6 +370,15 @@ Loop:
 					}
 				}
 			}
+			// never exceed the allowed label count lenght (63) 
+			if labels >= 63 { 
+				return "", lenmsg, &Error{err: "name exceeds 63 labels"}
```
Comment (gibson042) [inline msg.go]: There's no such limit in DNS; [RFC 1035](https://tools.ietf.org/html/rfc1035#section-2.3.4) requires the wire-format length of _each_ label to be 63 octets or less, which is already guaranteed by the `c & 0xC0` check.

File: msg.go [hunk 2]
```diff
@@ -369,6 +370,15 @@ Loop:
...
+			}
+			labels += 1
+			// never exceed the allowed doman name length (255 octets) 
+			if len(s) >= 255 {
```
Comment (gibson042) [inline msg.go]: This is erroneously checking presentation-format length, but the [RFC 1035 limit](https://tools.ietf.org/html/rfc1035#section-2.3.4) applies to _wire_ format. As a result, it rejects the following perfectly valid (if unusual) domain name:
```
\.\(\)\;\.\(\)\;\.\(\)\;.32-octets-including-length--001.32-octets-including-length--002.32-octets-including-length--003.32-octets-including-length--004.32-octets-including-length--005.32-octets-including-length--006.32-octets-including-length--007.13-octet-TLD.
```

Also, note that 255 octets is an acceptable length; it's 256 that is invalid.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] limiting domain nam..." ]
>gibson042 commented on this pull request.

Thanks for double checking this. I'll take a look.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] limiting domain nam..." ]
>gibson042 commented on this pull request.
>
>The checks in this merge are incorrect; it should be reverted.
>
>> @@ -369,6 +370,15 @@ Loop:
> 					}
> 				}
> 			}
>+			// never exceed the allowed label count lenght (63)
>+			if labels >= 63 {
>+				return "", lenmsg, &Error{err: "name exceeds 63 labels"}
>
>There's no such limit in DNS; [RFC 1035](https://tools.ietf.org/html/rfc1035#section-2.3.4) requires the wire-format length of _each_ label to be 63 octets or less, which is already guaranteed by the `c & 0xC0` check.
>
>> @@ -369,6 +370,15 @@ Loop:
> 					}
> 				}
> 			}
>+			// never exceed the allowed label count lenght (63)
>+			if labels >= 63 {
>+				return "", lenmsg, &Error{err: "name exceeds 63 labels"}
>+			}
>+			labels += 1
>+			// never exceed the allowed doman name length (255 octets)
>+			if len(s) >= 255 {
>
>This is erroneously checking presentation-format length, but the [RFC 1035 limit](https://tools.ietf.org/html/rfc1035#section-2.3.4) applies to _wire_ format. As a result, it rejects the following perfectly valid (if unusual) domain name:
>```
>\.\(\)\;\.\(\)\;\.\(\)\;.32-octets-including-length--001.32-octets-including-length--002.32-octets-including-length--003.32-octets-including-length--004.32-octets-including-length--005.32-octets-including-length--006.32-octets-including-length--007.13-octet-TLD.
>```
>
>Also, note that 255 octets is an acceptable length; it's 256 that is invalid.
>

Yes. :-( Care to send revert PR? We still have an issue here though.

**Comment by miekg** [comment]
@ogud any comments before stuff gets yanked out?

**Comment by ogud** [comment]
I'm working on a fix

Give me 24 hours, suffering from jet lag

Ólafur

On March 7, 2017 1:29:08 PM GMT+00:00, Miek Gieben <notifications@github.com> wrote:
>@ogud any comments before stuff gets yanked out?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/miekg/dns/pull/463#issuecomment-284720995

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] limiting domain nam..." ]
>I'm working on a fix
>
>Give me 24 hours, suffering from jet lag

Ack. Thanks

**Pull request merged**